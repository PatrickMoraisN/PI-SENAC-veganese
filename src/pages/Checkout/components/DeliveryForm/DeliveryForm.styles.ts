import styled from 'styled-components'

export const Form = styled.form`
  margin-top: 32px;
  display: flex;
  flex-direction: column;
  gap: 16px;

  input {
    background: #eeeded;
    border: 1px solid #e6e5e5;
    border-radius: 4px;
    padding: 12px;

    &::placeholder {
      color: #8d8686;
    }
  }
`

export const CepInput = styled.input`
  width: 200px;
`

export const CepError = styled.span`
  color: red;
`

export const InputWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  gap: 12px;
`

export const ComplementINput = styled.input`
  width: 100%;
`

export const UFInput = styled.input`
  cursor: not-allowed;
  width: 60px;
`

export const CityInput = styled.input`
  cursor: not-allowed;
  width: 276px;
`
