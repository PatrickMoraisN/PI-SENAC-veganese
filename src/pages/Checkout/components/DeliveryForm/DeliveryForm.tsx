import React, {
  ChangeEventHandler,
  useContext,
  useEffect,
  useState,
} from 'react'
import * as S from './DeliveryForm.styles'
import CepAPI from '@/api/cep'
import { UserInfoContext } from '@/contexts/UserInfoContext'

type CepInfoProps = {
  street: string
  city: string
  uf: string
  neighb: string
}

type DeliveryFormProps = {
  setDeliveryFormCompleted: (value: boolean) => void
  setError: (value: null) => void
}

interface EventProps extends ChangeEventHandler {
  target: HTMLInputElement
}

export function DeliveryForm({
  setDeliveryFormCompleted,
  setError,
}: DeliveryFormProps) {
  const [cepInfo, setCepInfo] = useState<null | CepInfoProps>(null)
  const [numberStreet, setNumberStreet] = useState('')
  const [complement, setComplement] = useState('')
  const [cepError, setCepError] = useState<null | string>(null)
  const { setUserInfo } = useContext(UserInfoContext)

  const formCompleted =
    cepInfo?.city && cepInfo.neighb && cepInfo.street && cepInfo.uf

  const getInfos = async (cep: string) => {
    try {
      if (cepError) {
        setCepError(null)
      }
      if (!cepInfo) {
        const cepResponse = await CepAPI.getCep(cep)
        setCepInfo({
          street: cepResponse.logradouro,
          city: cepResponse.localidade,
          uf: cepResponse.uf,
          neighb: cepResponse.bairro,
        })
      }
    } catch {
      setCepError('O CEP informado está incorreto! Informe um CEP válido.')
    }
  }

  const handleNumberChange = (e: EventProps) => {
    setError(null)
    setNumberStreet(e.target.value)
  }

  const handleComplementChange = (e: EventProps) => {
    setError(null)
    setComplement(e.target.value)
  }

  useEffect(() => {
    if (formCompleted) {
      const { street, city, uf, neighb } = cepInfo
      setDeliveryFormCompleted(true)
      setUserInfo((state) => ({ ...state, street, city, uf, neighb }))
    }
  }, [cepInfo])

  return (
    <>
      <h3>Endereço de Entrega</h3>
      <p>Informe o endereço onde deseja receber seu pedido</p>

      <S.Form>
        <S.CepInput
          type="number"
          placeholder="CEP"
          data-testid="cep"
          onBlur={(e) => getInfos((e.target as HTMLInputElement).value)}
        />
        {cepError && <S.CepError>{cepError}</S.CepError>}

        <input
          type="text"
          style={{ cursor: 'not-allowed' }}
          placeholder="Rua"
          disabled
          value={cepInfo ? cepInfo.street : ''}
          data-testid="street"
        />

        <S.InputWrapper>
          <S.CepInput
            type="number"
            placeholder="Número"
            onChange={handleNumberChange}
            value={numberStreet}
          />
          <S.ComplementINput
            type="text"
            placeholder="Complemento"
            value={complement}
            onChange={handleComplementChange}
          />
        </S.InputWrapper>

        <S.InputWrapper>
          <S.CepInput
            type="text"
            placeholder="Bairro"
            style={{ cursor: 'not-allowed' }}
            disabled
            value={cepInfo ? cepInfo.neighb : ''}
            data-testid="neighb"
          />
          <S.CityInput
            type="text"
            placeholder="Cidade"
            disabled
            value={cepInfo ? cepInfo.city : ''}
            data-testid="city"
          />
          <S.UFInput
            type="text"
            placeholder="UF"
            disabled
            value={cepInfo ? cepInfo.uf : ''}
            data-testid="uf"
          />
        </S.InputWrapper>
      </S.Form>
    </>
  )
}
