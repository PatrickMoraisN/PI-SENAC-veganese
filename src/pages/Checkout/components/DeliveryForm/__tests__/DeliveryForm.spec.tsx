import CepAPI from '@/api/cep'
import userEvent from '@testing-library/user-event'
import { render, screen } from '@testing-library/react'
import { DeliveryForm } from '../DeliveryForm'
import { RenderWithContext } from '@/test/util/RenderWithContext'

const setDeliveryFormCompletedMock = jest.fn()
const setErrorMock = jest.fn()

const CepAPIResponseMock = {
  logradouro: 'logradouroMock',
  localidade: 'localidadeMock',
  uf: 'ufMock',
  bairro: 'bairroMock',
}

const renderDeliveryForm = () =>
  render(
    <RenderWithContext>
      <DeliveryForm
        setDeliveryFormCompleted={setDeliveryFormCompletedMock}
        setError={setErrorMock}
      />
    </RenderWithContext>,
  )

jest.spyOn(CepAPI, 'getCep').mockResolvedValue(CepAPIResponseMock)

describe('DeliveryForm', () => {
  it('should render correctly', () => {
    const { container } = renderDeliveryForm()

    expect(container).toMatchSnapshot()
  })

  it('should call CepAPI correctly', async () => {
    renderDeliveryForm()

    await userEvent.type(screen.getByTestId('cep'), '22012013')

    const street = screen.getByTestId('street')
    await userEvent.click(street)

    expect(street).toHaveValue(CepAPIResponseMock.logradouro)
    expect(screen.getByTestId('neighb')).toHaveValue(CepAPIResponseMock.bairro)
    expect(screen.getByTestId('city')).toHaveValue(
      CepAPIResponseMock.localidade,
    )
    expect(screen.getByTestId('uf')).toHaveValue(CepAPIResponseMock.uf)
  })
})
