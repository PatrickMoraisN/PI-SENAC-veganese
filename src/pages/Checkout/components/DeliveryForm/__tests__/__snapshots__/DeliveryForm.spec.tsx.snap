// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeliveryForm should render correctly 1`] = `
<div>
  <h3>
    Endereço de Entrega
  </h3>
  <p>
    Informe o endereço onde deseja receber seu pedido
  </p>
  <form
    class="sc-bcXHqe bIkZqB"
  >
    <input
      class="sc-gswNZR cmQhvm"
      data-testid="cep"
      placeholder="CEP"
      type="number"
    />
    <input
      data-testid="street"
      disabled=""
      placeholder="Rua"
      style="cursor: not-allowed;"
      type="text"
      value=""
    />
    <div
      class="sc-hLBbgP iQhqPF"
    >
      <input
        class="sc-gswNZR cmQhvm"
        placeholder="Número"
        type="number"
        value=""
      />
      <input
        class="sc-eDvSVe bJxIbO"
        placeholder="Complemento"
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-hLBbgP iQhqPF"
    >
      <input
        class="sc-gswNZR cmQhvm"
        data-testid="neighb"
        disabled=""
        placeholder="Bairro"
        style="cursor: not-allowed;"
        type="text"
        value=""
      />
      <input
        class="sc-gKPRtg gMVRru"
        data-testid="city"
        disabled=""
        placeholder="Cidade"
        type="text"
        value=""
      />
      <input
        class="sc-jSUZER iNWTAz"
        data-testid="uf"
        disabled=""
        placeholder="UF"
        type="text"
        value=""
      />
    </div>
  </form>
</div>
`;
